lib/Analysis/NoiseAnalysis/BGV/NoiseByVarianceCanEmbModel.cpp

```diff
@@ -0,0 +1,300 @@
+#include "lib/Analysis/NoiseAnalysis/BGV/NoiseByVarianceCanEmbModel.h"
+
+#include <algorithm>
+#include <cassert>
+#include <cmath>
+#include <iomanip>
+#include <ios>
+#include <numeric>
+#include <sstream>
+#include <string>
+
+namespace mlir {
+namespace heir {
+namespace bgv {
+
+// the formulae below are mainly taken from MMLGA22
+// "Finding and Evaluating Parameters for BGV"
+// https://eprint.iacr.org/2022/706
+
+template <bool P>
+using Model = NoiseByVarianceCanEmbModel<P>;
+
```

Comment:
IIRC MMLGA22 tracks `||a||^{can}` of a ciphertext, which should be a bound. I would suggest a name like `NoiseCanEmbModel` or `NoiseByBoundCanEmbModel`.

In the code it seems that `StateType` stores the bound, though I admit during bound evolution a lot of variance are involved like `BScale` and `BKs`.

I call the model of MP24 by the name `NoiseByVarianceCoeffModel` because it tracks `Var(a)` instead of a bound `||a||`. In my mind `NoiseByVarinaceCanonModel` should track `Var(a^{can})` rather than `||a^{can}||`, though indeed little recent literature studies `Var(a^{can})`.

Reply:
Also please note that `toLogBound` is meant for `||a||`. So we need to reduce `||a||^{can}` to `||a||`. Fortunately for power-of-two 2 we have `||a|| <= c_m ||a||^{can}` and `c_m = 1`, so we can just return the value in StateType.

It would be better if we add a comment here saying because `||a|| <= ||a||^{can}` the bound on canonical embedding can be directly returned for use as coefficient embedding bound.

Reply:
Should we keep track of the ring expansion factor c_m, in case that we are not dealing with a power-of-two ring?

Reply:
That would be better. We might have a `getRingExpansionFactor`, which asserts on `getRingDim is power of 2` and returns `c_m = 1`. Then future development on non-power-of-two case, if any, would hit this assertion and know what to modify.

Reply:
I renamed the mode `NoiseCanEmbModel` as you suggested, added the comment and a `getRingExpansionFactor` function that asserts if the ring dimension is a power-of-two. I'm sorry, but somehow I lost the fight with git/hub and you can't really see the changes in `NoiseCanEmbModel.cpp`, because git thinks it is a new file.